File: vibeshield-ts/src/core/rce-detector.ts
export interface Finding {
  line: number;
  pattern: string;
  severity: 'critical' | 'high' | 'warning';
  suggestion: string;
  originalCode: string;
}

export interface DetectionResult {
  safe: boolean;
  findings: Finding[];
}

/**
 * State machine that masks comment bodies and string literals into spaces,
 * keeping length, index, newlines, and template literal interpolations intact.
 *
 * @param code - The raw source code to mask.
 * @returns The masked source code.
 */
function maskCommentsAndStrings(code: string): string {
  const chars = code.split('');
  const masked: string[] = [];
  let i = 0;
  
  const stateStack: string[] = ['NORMAL'];
  const braceStack: number[] = [];
  
  while (i < chars.length) {
    const currentState = stateStack[stateStack.length - 1];
    const c = chars[i];
    const next = chars[i + 1];
    
    if (currentState === 'NORMAL' || currentState === 'TEMPLATE_INTERPOLATION') {
      if (c === '/' && next === '/') {
        masked.push('/', '/');
        i += 2;
        while (i < chars.length && chars[i] !== '\n' && chars[i] !== '\r') {
          masked.push(' ');
          i++;
        }
        continue;
      }
      if (c === '/' && next === '*') {
        masked.push('/', '*');
        i += 2;
        while (i < chars.length && !(chars[i] === '*' && chars[i + 1] === '/')) {
          const char = chars[i];
          if (char === '\n' || char === '\r') {
            masked.push(char);
          } else {
            masked.push(' ');
          }
          i++;
        }
        if (i < chars.length) {
          masked.push('*', '/');
          i += 2;
        }
        continue;
      }
      if (c === "'") {
        stateStack.push('STRING_SINGLE');
        masked.push(c);
        i++;
        continue;
      }
      if (c === '"') {
        stateStack.push('STRING_DOUBLE');
        masked.push(c);
        i++;
        continue;
      }
      if (c === '`') {
        stateStack.push('STRING_TEMPLATE');
        masked.push(c);
        i++;
        continue;
      }
      if (c === '}' && stateStack.includes('TEMPLATE_INTERPOLATION')) {
        if (braceStack[braceStack.length - 1] === 0) {
          stateStack.pop();
          braceStack.pop();
          masked.push(c);
          i++;
          continue;
        } else {
          braceStack[braceStack.length - 1]--;
          masked.push(c);
          i++;
          continue;
        }
      }
      if (c === '{' && stateStack.includes('TEMPLATE_INTERPOLATION')) {
        braceStack[braceStack.length - 1]++;
        masked.push(c);
        i++;
        continue;
      }
      
      masked.push(c);
      i++;
    } else if (currentState === 'STRING_SINGLE') {
      if (c === '\\') {
        masked.push(' ', ' ');
        i += 2;
      } else if (c === "'") {
        stateStack.pop();
        masked.push(c);
        i++;
      } else if (c === '\n' || c === '\r') {
        masked.push(c);
        i++;
      } else {
        masked.push(' ');
        i++;
      }
    } else if (currentState === 'STRING_DOUBLE') {
      if (c === '\\') {
        masked.push(' ', ' ');
        i += 2;
      } else if (c === '"') {
        stateStack.pop();
        masked.push(c);
        i++;
      } else if (c === '\n' || c === '\r') {
        masked.push(c);
        i++;
      } else {
        masked.push(' ');
        i++;
      }
    } else if (currentState === 'STRING_TEMPLATE') {
      if (c === '\\') {
        masked.push(' ', ' ');
        i += 2;
      } else if (c === '$' && next === '{') {
        stateStack.push('TEMPLATE_INTERPOLATION');
        braceStack.push(0);
        masked.push('$', '{');
        i += 2;
      } else if (c === '`') {
        stateStack.pop();
        masked.push(c);
        i++;
      } else if (c === '\n' || c === '\r') {
        masked.push(c);
        i++;
      } else {
        masked.push(' ');
        i++;
      }
    }
  }
  return masked.join('');
}

/**
 * Parses function arguments matching opening/closing parentheses.
 *
 * @param code - The source code.
 * @param startIndex - The start index of the call statement.
 * @returns Object with full call, args string, and end index, or null.
 */
function extractCallArgs(code: string, startIndex: number): { fullCall: string; args: string; endIndex: number } | null {
  let openParens = 0;
  let i = startIndex;
  while (i < code.length && code[i] !== '(') {
    i++;
  }
  if (i >= code.length) return null;
  const argStart = i + 1;
  openParens = 1;
  i++;
  while (i < code.length && openParens > 0) {
    if (code[i] === '(') openParens++;
    else if (code[i] === ')') openParens--;
    if (openParens === 0) break;
    i++;
  }
  if (openParens > 0) return null;
  return {
    fullCall: code.substring(startIndex, i + 1),
    args: code.substring(argStart, i),
    endIndex: i
  };
}

/**
 * Returns the 1-based line number for a given character index.
 *
 * @param code - The source code.
 * @param index - Character index.
 * @returns 1-based line number.
 */
function getLineNumber(code: string, index: number): number {
  return code.substring(0, index).split('\n').length;
}

/**
 * Replaces string content inside single, double, backtick quotes with '***' to mask secrets.
 *
 * @param line - The original line of code.
 * @returns The masked line of code.
 */
function maskOriginalCode(line: string): string {
  return line
    .replace(/(["'])(?:(?=(\\?))\2.)*?\1/g, '$1***$1')
    .replace(/(`)(?:(?=(\\?))\2.)*?`/g, '$1***$1');
}

/**
 * Checks if the extracted argument expression is dynamic.
 *
 * @param args - The arguments string.
 * @returns True if the argument is dynamic, false otherwise.
 */
function isDynamic(args: string): boolean {
  const trimmed = args.trim();
  if (!trimmed) return false;
  // If it's a simple single string literal, e.g. "foo", 'foo', `foo`
  // After masking, the contents inside quotes are just spaces.
  // Check if it matches exactly a single quoted/backticked string consisting only of spaces/empty
  const staticPattern = /^(['"`])\s*\1$/;
  if (staticPattern.test(trimmed)) return false;
  // If it contains a plus sign outside quotes or a variable, it is dynamic.
  // In masked code, any non-space/non-operator word character outside quotes indicates a variable.
  // We can check if there are variable names or interpolation syntax
  if (trimmed.includes('+') || trimmed.includes('${')) return true;
  // Check for any unquoted identifier word character (excluding numeric constants)
  const clean = trimmed.replace(/['"`]\s*['"`]/g, '');
  if (/[a-zA-Z_$][a-zA-Z0-9_$]*/.test(clean)) return true;
  return false;
}

/**
 * Scans code for Remote Code Execution (RCE) patterns.
 *
 * @param code - The source code to scan.
 * @returns DetectionResult with safety flag and details of findings.
 */
export function detectRcePatterns(code: string): DetectionResult {
  const findings: Finding[] = [];
  if (!code) return { safe: true, findings };

  const masked = maskCommentsAndStrings(code);
  const lines = code.split('\n');

  // Helper to add findings
  const addFinding = (index: number, pattern: string, severity: 'critical' | 'high' | 'warning', suggestion: string) => {
    const lineNum = getLineNumber(code, index);
    const originalLine = lines[lineNum - 1] || '';
    findings.push({
      line: lineNum,
      pattern,
      severity,
      suggestion,
      originalCode: maskOriginalCode(originalLine.trim())
    });
  };

  // Rule 1: eval(...)
  const evalRegex = /\beval\s*(?=\()/g;
  let match;
  while ((match = evalRegex.exec(masked)) !== null) {
    const call = extractCallArgs(masked, match.index);
    if (call) {
      const dynamic = isDynamic(call.args);
      addFinding(
        match.index,
        'eval()',
        dynamic ? 'critical' : 'high',
        dynamic 
          ? 'Avoid dynamic code evaluation with eval(). Use safe parsing or explicit function mappings instead.'
          : 'eval() is highly discouraged. Refactor to use static code structures.'
      );
    }
  }

  // Rule 2: Function constructor / new Function
  const funcRegex = /\b(new\s+)?Function\s*(?=\()/g;
  while ((match = funcRegex.exec(masked)) !== null) {
    const call = extractCallArgs(masked, match.index);
    if (call) {
      const dynamic = isDynamic(call.args);
      addFinding(
        match.index,
        'Function constructor',
        dynamic ? 'critical' : 'high',
        dynamic
          ? 'Avoid dynamic code generation via Function constructor.'
          : 'Function constructor is discouraged. Use standard function declarations instead.'
      );
    }
  }

  // Rule 3: setTimeout/setInterval with string arg
  const timerRegex = /\b(setTimeout|setInterval)\s*(?=\()/g;
  while ((match = timerRegex.exec(masked)) !== null) {
    const call = extractCallArgs(masked, match.index);
    if (call) {
      // Get the first argument
      const firstArg = call.args.split(',')[0].trim();
      const startsWithQuote = /^['"`]/.test(firstArg);
      const dynamic = isDynamic(firstArg);
      const isArrowOrFunction = firstArg.includes('=>') || /^\bfunction\b/.test(firstArg);
      if (!isArrowOrFunction && (startsWithQuote || dynamic)) {
        addFinding(
          match.index,
          `${match[1]}() with string argument`,
          dynamic ? 'critical' : 'high',
          'Pass a function reference or arrow function to setTimeout/setInterval instead of a string literal.'
        );
      }
    }
  }

  // Rule 4: child_process methods
  const childProcRegex = /\b(child_process\.)?(exec|execSync|spawn|spawnSync|execFile|execFileSync)\s*(?=\()/g;
  while ((match = childProcRegex.exec(masked)) !== null) {
    const call = extractCallArgs(masked, match.index);
    if (call) {
      const firstArg = call.args.split(',')[0].trim();
      const dynamic = isDynamic(firstArg);
      
      // Also check if options object has shell: true
      const hasShellTrue = /\bshell\s*:\s*true\b/i.test(call.args);

      if (dynamic || hasShellTrue) {
        addFinding(
          match.index,
          `child_process.${match[2]}() with dynamic input`,
          'critical',
          'Never execute commands constructed from dynamic user inputs. Use spawn/execFile with an array of arguments, and ensure input is validated.'
        );
      } else {
        addFinding(
          match.index,
          `child_process.${match[2]}()`,
          'warning',
          'Ensure command execution parameters are static and do not accept untrusted input.'
        );
      }
    }
  }

  // Rule 5: vm methods
  const vmRegex = /\b(vm\.)?(runInNewContext|runInContext|runInThisContext|Script)\s*(?=\()/g;
  while ((match = vmRegex.exec(masked)) !== null) {
    const call = extractCallArgs(masked, match.index);
    if (call) {
      const firstArg = call.args.split(',')[0].trim();
      const dynamic = isDynamic(firstArg);
      addFinding(
        match.index,
        `vm.${match[2]}()`,
        dynamic ? 'critical' : 'high',
        dynamic
          ? 'Avoid dynamic code execution in VM context with untrusted inputs.'
          : 'VM code execution should be restricted to trusted templates.'
      );
    }
  }

  // Rule 6: Sandbox escapes (process.mainModule / process.binding)
  const escapeRegex = /\bprocess\.(mainModule|binding|_getActiveRequests|_getActiveHandles)\b/g;
  while ((match = escapeRegex.exec(masked)) !== null) {
    addFinding(
      match.index,
      `process.${match[1]}`,
      'critical',
      'Access to process internals is a security risk and indicates a sandbox escape attempt.'
    );
  }

  // Rule 7: Dynamic require/import
  const importRegex = /\b(require|import)\s*(?=\()/g;
  while ((match = importRegex.exec(masked)) !== null) {
    const call = extractCallArgs(masked, match.index);
    if (call) {
      const dynamic = isDynamic(call.args);
      if (dynamic) {
        addFinding(
          match.index,
          `dynamic ${match[1]}()`,
          'high',
          'Avoid dynamic require/import statements. Use static imports or a whitelist of allowed modules.'
        );
      }
    }
  }

  // Sort findings by line number
  findings.sort((a, b) => a.line - b.line);

  return {
    safe: findings.length === 0,
    findings
  };
}

File: vibeshield-python/src/rce_detector.py
import re
import logging
from typing import Any, Dict, List, Optional

logger = logging.getLogger("VibeShield")


def mask_comments_and_strings(code: str) -> str:
    """
    Masks Python comments and string literals in the code to assist pattern analysis.

    Supports single-line comments (#), single-line strings (', "), triple-quoted
    strings (''', \"\"\"[triple-double]), and f-string interpolation blocks ({...}).

    Args:
        code: The source code string to mask.

    Returns:
        A masked version of the source code where string literals and comments
        are replaced with spaces, preserving layout and f-string interpolation blocks.
    """
    chars = list(code)
    masked = []
    i = 0
    n = len(chars)
    state_stack = ['NORMAL']
    brace_stack = []

    while i < n:
        current_state = state_stack[-1]
        c = chars[i]
        next_c = chars[i+1] if i + 1 < n else ""
        next_2c = chars[i+2] if i + 2 < n else ""

        if current_state in ('NORMAL', 'INTERPOLATION'):
            if c == '#':
                masked.append('#')
                i += 1
                while i < n and chars[i] not in ('\n', '\r'):
                    masked.append(' ')
                    i += 1
                continue
            if c == '"' and next_c == '"' and next_2c == '"':
                is_f = i > 0 and chars[i-1].lower() == 'f'
                if is_f:
                    masked[-1] = ' '
                state_stack.append('TRIPLE_DOUBLE_F' if is_f else 'TRIPLE_DOUBLE')
                masked.extend(['"', '"', '"'])
                i += 3
                continue
            if c == "'" and next_c == "'" and next_2c == "'":
                is_f = i > 0 and chars[i-1].lower() == 'f'
                if is_f:
                    masked[-1] = ' '
                state_stack.append('TRIPLE_SINGLE_F' if is_f else 'TRIPLE_SINGLE')
                masked.extend(["'", "'", "'"])
                i += 3
                continue
            if c == '"':
                is_f = i > 0 and chars[i-1].lower() == 'f'
                if is_f:
                    masked[-1] = ' '
                state_stack.append('STRING_DOUBLE_F' if is_f else 'STRING_DOUBLE')
                masked.append('"')
                i += 1
                continue
            if c == "'":
                is_f = i > 0 and chars[i-1].lower() == 'f'
                if is_f:
                    masked[-1] = ' '
                state_stack.append('STRING_SINGLE_F' if is_f else 'STRING_SINGLE')
                masked.append("'")
                i += 1
                continue
            if state_stack.count('INTERPOLATION') > 0:
                if c == '{':
                    brace_stack[-1] += 1
                    masked.append(c)
                    i += 1
                    continue
                if c == '}':
                    if brace_stack[-1] == 0:
                        state_stack.pop()
                        brace_stack.pop()
                        masked.append(c)
                        i += 1
                        continue
                    else:
                        brace_stack[-1] -= 1
                        masked.append(c)
                        i += 1
                        continue
            masked.append(c)
            i += 1
        elif current_state == 'STRING_SINGLE':
            if c == '\\':
                masked.extend([' ', ' '])
                i += 2
            elif c == "'":
                state_stack.pop()
                masked.append(c)
                i += 1
            elif c in ('\n', '\r'):
                masked.append(c)
                i += 1
            else:
                masked.append(' ')
                i += 1
        elif current_state == 'STRING_DOUBLE':
            if c == '\\':
                masked.extend([' ', ' '])
                i += 2
            elif c == '"':
                state_stack.pop()
                masked.append(c)
                i += 1
            elif c in ('\n', '\r'):
                masked.append(c)
                i += 1
            else:
                masked.append(' ')
                i += 1
        elif current_state == 'TRIPLE_SINGLE':
            if c == '\\':
                masked.extend([' ', ' '])
                i += 2
            elif c == "'" and next_c == "'" and next_2c == "'":
                state_stack.pop()
                masked.extend(["'", "'", "'"])
                i += 3
            elif c in ('\n', '\r'):
                masked.append(c)
                i += 1
            else:
                masked.append(' ')
                i += 1
        elif current_state == 'TRIPLE_DOUBLE':
            if c == '\\':
                masked.extend([' ', ' '])
                i += 2
            elif c == '"' and next_c == '"' and next_2c == '"':
                state_stack.pop()
                masked.extend(['"', '"', '"'])
                i += 3
            elif c in ('\n', '\r'):
                masked.append(c)
                i += 1
            else:
                masked.append(' ')
                i += 1
        elif current_state in ('STRING_SINGLE_F', 'STRING_DOUBLE_F', 'TRIPLE_SINGLE_F', 'TRIPLE_DOUBLE_F'):
            is_triple = 'TRIPLE' in current_state
            quote_char = "'" if 'SINGLE' in current_state else '"'
            if c == '\\':
                masked.extend([' ', ' '])
                i += 2
            elif c == '{' and next_c != '{':
                state_stack.append('INTERPOLATION')
                brace_stack.append(0)
                masked.append('{')
                i += 1
            elif c == '{' and next_c == '{':
                masked.extend([' ', ' '])
                i += 2
            elif c == '}' and next_c == '}':
                masked.extend([' ', ' '])
                i += 2
            elif is_triple and c == quote_char and next_c == quote_char and next_2c == quote_char:
                state_stack.pop()
                masked.extend([quote_char, quote_char, quote_char])
                i += 3
            elif not is_triple and c == quote_char:
                state_stack.pop()
                masked.append(quote_char)
                i += 1
            elif c in ('\n', '\r'):
                masked.append(c)
                i += 1
            else:
                masked.append(' ')
                i += 1
    return "".join(masked)


def _extract_call_args(code: str, start_index: int) -> Optional[Dict[str, Any]]:
    """
    Extracts call arguments matching parentheses starting from start_index.

    Args:
        code: The masked or original source code.
        start_index: The starting character index of the function call name.

    Returns:
        A dictionary with keys 'full_call', 'args', and 'end_index', or None if unmatched.
    """
    open_parens = 0
    i = start_index
    n = len(code)
    while i < n and code[i] != '(':
        i += 1
    if i >= n:
        return None
    arg_start = i + 1
    open_parens = 1
    i += 1
    while i < n and open_parens > 0:
        if code[i] == '(':
            open_parens += 1
        elif code[i] == ')':
            open_parens -= 1
        if open_parens == 0:
            break
        i += 1
    if open_parens > 0:
        return None
    return {
        'full_call': code[start_index:i+1],
        'args': code[arg_start:i],
        'end_index': i
    }


def _get_line_number(code: str, index: int) -> int:
    """
    Computes the 1-based line number for a given character index in the code.

    Args:
        code: The original source code.
        index: The character index.

    Returns:
        The 1-based line number.
    """
    return len(code[:index].split('\n'))


def _mask_original_code(line: str) -> str:
    """
    Masks string literal contents with '***' while keeping f-string interpolation and structure.

    Args:
        line: A line of original source code.

    Returns:
        The line with string contents masked.
    """
    chars = list(line)
    n = len(chars)
    masked = []
    i = 0
    state_stack = ['NORMAL']
    brace_stack = []

    while i < n:
        current_state = state_stack[-1]
        c = chars[i]
        next_c = chars[i+1] if i + 1 < n else ""
        next_2c = chars[i+2] if i + 2 < n else ""

        if current_state in ('NORMAL', 'INTERPOLATION'):
            if c == '#':
                masked.append('#')
                masked.extend([' '] * (n - i - 1))
                break
            if c == '"' and next_c == '"' and next_2c == '"':
                is_f = i > 0 and chars[i-1].lower() == 'f'
                state_stack.append('TRIPLE_DOUBLE_F' if is_f else 'TRIPLE_DOUBLE')
                masked.extend(['"', '"', '"'])
                if not is_f:
                    masked.append('***')
                i += 3
                continue
            if c == "'" and next_c == "'" and next_2c == "'":
                is_f = i > 0 and chars[i-1].lower() == 'f'
                state_stack.append('TRIPLE_SINGLE_F' if is_f else 'TRIPLE_SINGLE')
                masked.extend(["'", "'", "'"])
                if not is_f:
                    masked.append('***')
                i += 3
                continue
            if c == '"':
                is_f = i > 0 and chars[i-1].lower() == 'f'
                state_stack.append('STRING_DOUBLE_F' if is_f else 'STRING_DOUBLE')
                masked.append('"')
                if not is_f:
                    masked.append('***')
                i += 1
                continue
            if c == "'":
                is_f = i > 0 and chars[i-1].lower() == 'f'
                state_stack.append('STRING_SINGLE_F' if is_f else 'STRING_SINGLE')
                masked.append("'")
                if not is_f:
                    masked.append('***')
                i += 1
                continue
            if state_stack.count('INTERPOLATION') > 0:
                if c == '{':
                    brace_stack[-1] += 1
                    masked.append(c)
                    i += 1
                    continue
                if c == '}':
                    if brace_stack[-1] == 0:
                        state_stack.pop()
                        brace_stack.pop()
                        masked.append(c)
                        i += 1
                        continue
                    else:
                        brace_stack[-1] -= 1
                        masked.append(c)
                        i += 1
                        continue
            masked.append(c)
            i += 1
        elif current_state in ('STRING_SINGLE', 'STRING_DOUBLE', 'TRIPLE_SINGLE', 'TRIPLE_DOUBLE'):
            is_triple = 'TRIPLE' in current_state
            quote_char = "'" if 'SINGLE' in current_state else '"'
            if c == '\\':
                i += 2
            elif is_triple and c == quote_char and next_c == quote_char and next_2c == quote_char:
                state_stack.pop()
                masked.extend([quote_char, quote_char, quote_char])
                i += 3
            elif not is_triple and c == quote_char:
                state_stack.pop()
                masked.append(quote_char)
                i += 1
            else:
                i += 1
        elif current_state in ('STRING_SINGLE_F', 'STRING_DOUBLE_F', 'TRIPLE_SINGLE_F', 'TRIPLE_DOUBLE_F'):
            is_triple = 'TRIPLE' in current_state
            quote_char = "'" if 'SINGLE' in current_state else '"'
            if c == '\\':
                masked.extend([' ', ' '])
                i += 2
            elif c == '{' and next_c != '{':
                state_stack.append('INTERPOLATION')
                brace_stack.append(0)
                masked.append('{')
                i += 1
            elif c == '{' and next_c == '{':
                masked.extend(['*', '*'])
                i += 2
            elif c == '}' and next_c == '}':
                masked.extend(['*', '*'])
                i += 2
            elif is_triple and c == quote_char and next_c == quote_char and next_2c == quote_char:
                state_stack.pop()
                masked.extend([quote_char, quote_char, quote_char])
                i += 3
            elif not is_triple and c == quote_char:
                state_stack.pop()
                masked.append(quote_char)
                i += 1
            elif c in ('\n', '\r'):
                masked.append(c)
                i += 1
            else:
                masked.append('*')
                i += 1

    return "".join(masked)


def _is_dynamic(args: str) -> bool:
    """
    Checks if the function arguments contain dynamic operations.

    Args:
        args: The string of extracted arguments.

    Returns:
        True if the arguments contain variable names, concatenations, or formatting.
    """
    trimmed = args.strip()
    if not trimmed:
        return False
    # If it is a simple single string literal (after masking, has space or empty inside matching quotes)
    static_patterns = [
        r"^(['\"])\s*\1$",
        r"^([\"']{3})\s*\1$"
    ]
    if any(re.match(p, trimmed) for p in static_patterns):
        return False
    # If it contains variable lookup, format operations or concatenations
    if '+' in trimmed or '{' in trimmed or '%' in trimmed:
        return True
    # Strip quotes
    clean = re.sub(r'([\'"])\s*\1', '', trimmed)
    clean = re.sub(r'([\'"]{3})\s*\1', '', clean)
    if re.search(r'[a-zA-Z_][a-zA-Z0-9_]*', clean):
        return True
    return False


def _get_positional_args(args: str) -> List[str]:
    """
    Parses the arguments string into individual arguments, filtering out keyword arguments.

    Args:
        args: The string of arguments inside call parentheses.

    Returns:
        A list of positional argument strings.
    """
    i = 0
    n = len(args)
    parens = 0
    brackets = 0
    braces = 0
    arg_start = 0
    parts = []
    while i < n:
        c = args[i]
        if c == '(':
            parens += 1
        elif c == ')':
            parens -= 1
        elif c == '[':
            brackets += 1
        elif c == ']':
            brackets -= 1
        elif c == '{':
            braces += 1
        elif c == '}':
            braces -= 1
        elif c == ',' and parens == 0 and brackets == 0 and braces == 0:
            parts.append(args[arg_start:i].strip())
            arg_start = i + 1
        i += 1
    if arg_start < n:
        parts.append(args[arg_start:].strip())

    positional = []
    for part in parts:
        if not part:
            continue
        # Filter out keyword arguments (e.g., shell=True)
        if re.match(r'^[a-zA-Z_][a-zA-Z0-9_]*\s*=(?!=)', part):
            continue
        positional.append(part)
    return positional


def detect_rce_patterns(code: str) -> Dict[str, Any]:
    """
    Analyzes Python source code for Remote Code Execution (RCE) patterns.

    Detects eval, exec, compile, dangerous os/subprocess usage, and sandbox escapes.

    Args:
        code: The Python source code to analyze.

    Returns:
        A dictionary containing "safe" (bool) and "findings" (list of dicts).
    """
    findings = []
    if not code:
        return {"safe": True, "findings": findings}

    masked = mask_comments_and_strings(code)
    lines = code.split('\n')

    def add_finding(index: int, pattern: str, severity: str, suggestion: str):
        line_num = _get_line_number(code, index)
        original_line = lines[line_num - 1] if line_num - 1 < len(lines) else ''
        findings.append({
            "line": line_num,
            "pattern": pattern,
            "severity": severity,
            "suggestion": suggestion,
            "original_code": _mask_original_code(original_line.strip())
        })

    # Rule 1: eval(...) / exec(...) / compile(...)
    for func in ('eval', 'exec', 'compile'):
        pattern_regex = re.compile(rf'\b{func}\s*(?=\()')
        for match in pattern_regex.finditer(masked):
            call = _extract_call_args(masked, match.start())
            if call:
                pos_args = _get_positional_args(call['args'])
                dynamic = any(_is_dynamic(arg) for arg in pos_args) if pos_args else False
                add_finding(
                    match.start(),
                    f"{func}()",
                    "critical" if dynamic else "high",
                    f"Avoid using {func}(). Use safe parsers or standard operations instead."
                )

    # Rule 2: os.system / os.popen / os.spawn* / os.exec*
    os_funcs = (
        'system', 'popen', 'spawnl', 'spawnle', 'spawnlp', 'spawnlpe',
        'spawnv', 'spawnve', 'spawnvp', 'spawnvpe', 'execl', 'execle',
        'execlp', 'execlpe', 'execv', 'execve', 'execvp', 'execvpe'
    )
    for func in os_funcs:
        pattern_regex = re.compile(rf'\b(os\.)?{func}\s*(?=\()')
        for match in pattern_regex.finditer(masked):
            call = _extract_call_args(masked, match.start())
            if call:
                pos_args = _get_positional_args(call['args'])
                dynamic = any(_is_dynamic(arg) for arg in pos_args) if pos_args else False
                add_finding(
                    match.start(),
                    f"os.{func}()",
                    "critical" if dynamic else "high",
                    f"os.{func}() execution is unsafe. Use subprocess module with shell=False and argument arrays instead."
                )

    # Rule 3: subprocess
    sub_funcs = ('run', 'Popen', 'call', 'check_output', 'check_call')
    for func in sub_funcs:
        pattern_regex = re.compile(rf'\b(subprocess\.)?{func}\s*(?=\()')
        for match in pattern_regex.finditer(masked):
            call = _extract_call_args(masked, match.start())
            if call:
                pos_args = _get_positional_args(call['args'])
                dynamic = any(_is_dynamic(arg) for arg in pos_args) if pos_args else False
                has_shell_true = bool(re.search(r'\bshell\s*=\s*True\b', call['args']))
                if has_shell_true and dynamic:
                    add_finding(
                        match.start(),
                        f"subprocess.{func}(shell=True) with dynamic input",
                        "critical",
                        "CRITICAL: Running subprocess with shell=True and dynamic arguments is highly vulnerable to command injection. Avoid shell=True and pass command arguments as a list."
                    )
                elif has_shell_true:
                    add_finding(
                        match.start(),
                        f"subprocess.{func}(shell=True)",
                        "high",
                        "subprocess with shell=True is dangerous. Disable shell mode and pass arguments as a list."
                    )
                elif dynamic:
                    add_finding(
                        match.start(),
                        f"subprocess.{func}() with dynamic input",
                        "warning",
                        "Dynamic subprocess inputs should be sanitised and validated before execution."
                    )

    # Rule 4: Sandbox escapes (__import__, globals, locals)
    for func in ('__import__',):
        pattern_regex = re.compile(rf'\b{func}\s*(?=\()')
        for match in pattern_regex.finditer(masked):
            call = _extract_call_args(masked, match.start())
            if call:
                pos_args = _get_positional_args(call['args'])
                dynamic = any(_is_dynamic(arg) for arg in pos_args) if pos_args else False
                add_finding(
                    match.start(),
                    "__import__()",
                    "high" if dynamic else "warning",
                    "Avoid dynamic imports with __import__(). Use importlib instead."
                )

    for func in ('globals', 'locals'):
        pattern_regex = re.compile(rf'\b{func}\s*(?=\()')
        for match in pattern_regex.finditer(masked):
            add_finding(
                match.start(),
                f"{func}()",
                "warning",
                f"Avoid using {func}() to prevent dynamic variable access or modification."
            )

    # Rule 5: getattr/setattr dynamic attribute lookup on variables
    for func in ('getattr', 'setattr'):
        pattern_regex = re.compile(rf'\b{func}\s*(?=\()')
        for match in pattern_regex.finditer(masked):
            call = _extract_call_args(masked, match.start())
            if call:
                pos_args = _get_positional_args(call['args'])
                if len(pos_args) >= 2:
                    second_arg = pos_args[1]
                    if _is_dynamic(second_arg):
                        add_finding(
                            match.start(),
                            f"dynamic {func}() attribute lookup",
                            "high",
                            f"Dynamic attribute resolution using {func}() with variables can be abused. Restrict attribute names using a whitelist."
                        )

    findings.sort(key=lambda x: x['line'])
    return {"safe": len(findings) == 0, "findings": findings}
